src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -182,20 +183,34 @@ private static void AddDataProtection(OrchardCoreBuilder builder)
                 var settings = serviceProvider.GetRequiredService<ShellSettings>();
                 var options = serviceProvider.GetRequiredService<IOptions<ShellOptions>>();
 
-                var directory = Directory.CreateDirectory(Path.Combine(
-                options.Value.ShellsApplicationDataPath,
-                options.Value.ShellsContainerName,
-                settings.Name, "DataProtection-Keys"));
+                DirectoryInfo directory = null;
+
+                try
+                {
+                    directory = Directory.CreateDirectory(Path.Combine(
+                        options.Value.ShellsApplicationDataPath,
+                        options.Value.ShellsContainerName,
+                        settings.Name, "DataProtection-Keys"));
+                }
+                catch (UnauthorizedAccessException)
+                {
+                }
 
                 // Re-register the data protection services to be tenant-aware so that modules that internally
                 // rely on IDataProtector/IDataProtectionProvider automatically get an isolated instance that
                 // manages its own key ring and doesn't allow decrypting payloads encrypted by another tenant.
                 // By default, the key ring is stored in the tenant directory of the configured App_Data path.
-                var collection = new ServiceCollection()
+                var dpBuilder = new ServiceCollection()
                     .AddDataProtection()
                     .PersistKeysToFileSystem(directory)
-                    .SetApplicationName(settings.Name)
-                    .Services;
+                    .SetApplicationName(settings.Name);
+
+                if (directory != null)
+                {
+                    dpBuilder.PersistKeysToFileSystem(directory);
```

Comment:
The question is where is the thing stored? Also will it still be isolated by tenant?
I would have preferred a way to configure the location of the keys instead.

Reply:
It depends on the environment, see here

https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/default-settings?view=aspnetcore-2.1

Where they say

> The first three options above should provide good defaults

And also

> If none of these conditions match, keys aren't persisted outside of the current process. When the process shuts down, all generated keys are lost.

Maybe need to be confirmed by @PinpointTownes